### PR TITLE
Add support for new Ginlong/Solis Wifi stick (S3-WIFI-ST)

### LINKED ISSUE #289: Support for latest Ginlong/Solis wifi stick (S3-WIFI-ST)
Hi,

I recently got a Solis inverter and it seems to now come with a new Wifi stick (S3-WIFI-ST).
Unfortunately the web UI is slightly different here and although it's still a web site with authentication and a status.html page, the page now contains actual HTML instead of the vars expected for the old stick and the HTML is populated via JS:
```
var inverter = new XMLHttpRequest();
inverter.open("get", "inverter.cgi?t=" + Math.random(), true);
inverter.send();
inverter.onreadystatechange = function() {
  if (inverter.readyState == 4 && inverter.status == 200) {
    var inverterArr = inverter.responseText.split(";");
    set_inner_by_id("webdata_sn", inverterArr[0]);
    set_inner_by_id("webdata_msvn", inverterArr[1]);
    set_inner_by_id("webdata_pv_type", inverterArr[2]);
    set_inner_by_id("webdata_rate_p", inverterArr[3] + "℃");
    set_inner_by_id("webdata_now_p", inverterArr[4] + "W");
    set_inner_by_id("webdata_today_e", inverterArr[5] + "kWh");
    set_inner_by_id("webdata_total_e", inverterArr[6] + "kWh");
    set_inner_by_id("webdata_alarm", inverterArr[7]);
  } else {
    console.log(inverter.status)
  }
```
Accssing the cgi URL directly gives output such as:
`380XXX228080XXX;7800XX;205;44.9;4210;6.300000;d;NO; `

Seeing as this is now quite different from the original format and that the project was originally for omnik inverters and not solis, do you see any point in adding support for it as part of this project?
I can try whipping up an implementation, otherwise it might be better suited in its own project.

Cheers

### PR DESCRIPTION
**Pull request by koyote**
Fixes #289

It introduces a new source type called 'cgi' which can be used for the newer Wifi stick.

### CODE REVIEW

**Comment by MarijnS95** [comment]
@koyote if you replace the header with `Fixes #289` it'll auto-close the issue on successful merge!

File: omnikinverter/omnikinverter.py [hunk 1]
```diff
@@ -179,6 +179,9 @@ async def inverter(self) -> Inverter:
         if self.source_type == "html":
             data = await self.request("status.html")
             return Inverter.from_html(data)
+        if self.source_type == "cgi":
+            data = await self.request("inverter.cgi", params={"t": "123"})
```
Comment (MarijnS95) [inline omnikinverter/omnikinverter.py:183]: What's this `123` param for?

File: omnikinverter/models.py [hunk 2]
```diff
@@ -200,6 +200,40 @@ def from_tcp(data: dict[str, Any]) -> Inverter:
...
+
+        return Inverter(
+            serial_number=split_data[0],
+            model=split_data[2],
+            firmware=split_data[1],
+            firmware_slave=None,
+            solar_rated_power=None,
+            solar_current_power=try_parse_float(split_data[4]),
+            solar_energy_today=try_parse_float(split_data[5]),
+            solar_energy_total=try_parse_float(split_data[6]),
```
Comment (MarijnS95) [inline omnikinverter/models.py:234]: How do you know that this is encoded in the seventh field? Your test fixture has the non-float value `d` here?

File: tests/fixtures/inverter.cgi [hunk 3]
```diff
@@ -0,0 +1 @@
+12345678910;3280031;205;38.4;780;23.799999;d;NO;
```
Comment (MarijnS95) [inline tests/fixtures/inverter.cgi:1]: The fourth field (of value `38.4`) is not parsed anywhere. Any idea what it means, maybe inverter temperature (we have a field for that)?

File: omnikinverter/omnikinverter.py (hunk 1 above)
Comment (koyote) [reply omnikinverter/omnikinverter.py:183]: No idea :)
The Web UI sends the following request:
```
var inverter = new XMLHttpRequest();
inverter.open("get", "inverter.cgi?t=" + Math.random(), true);
```

I figured the value for t is not important so kept it constant.
I can change it to send in a random number to more closely follow the Web UI if you like.

**Comment by MarijnS95** [comment]
@koyote Not the title, the first sentence in your PR description (sorry for using the unclear word "header" here).

File: omnikinverter/models.py (hunk 2 above)
Comment (koyote) [reply omnikinverter/models.py:234]: I am going by the variable names the Web UI uses in its request:

```
    set_inner_by_id("webdata_sn", inverterArr[0]);
    set_inner_by_id("webdata_msvn", inverterArr[1]);
    set_inner_by_id("webdata_pv_type", inverterArr[2]);
    set_inner_by_id("webdata_rate_p", inverterArr[3] + "℃");
    set_inner_by_id("webdata_now_p", inverterArr[4] + "W");
    set_inner_by_id("webdata_today_e", inverterArr[5] + "kWh");
    set_inner_by_id("webdata_total_e", inverterArr[6] + "kWh");
    set_inner_by_id("webdata_alarm", inverterArr[7]);
```
They match with the 'old' ones.
The data on my Web UI displays 'd' for that field:
Screenshot_2022-11-04_19-31-46


That being said, I noticed that webdata_rate_p is wrong as it returns the temperature instead...

**Comment by MarijnS95** [comment]
Note that #247 is going to add support for `alarm_value`, which this will have to fill based on `set_inner_by_id("webdata_alarm", inverterArr[7]);` (CC @klaasnicolaas, pending which PR gets in first).

File: tests/fixtures/inverter.cgi (hunk 3 above)
Comment (koyote) [reply tests/fixtures/inverter.cgi:1]: It is indeed the temperature. I somehow missed that parameter as part of the TCP values.
I'll populate it.

File: omnikinverter/models.py (hunk 2 above)
Comment (MarijnS95) [reply omnikinverter/models.py:234]: Looks like they just reused an existing `id=` and changed the unit and label text (with a typo `temprature` :joy:) :)

**Comment by MarijnS95** [comment]
Don't forget to add your device to the `Supported models` list in the `README` with the new data type, and change the `On the tcp source type you can also find:` section to mention temperature on your CGI backend :)

(Although I do not know if this is specific to your device, since they abused/reused the `webdata_rate_p` field for it...)

**Comment by koyote** [comment]
I've added the temperature and updated the README

File: README.md [hunk 4]
```diff
@@ -93,9 +94,12 @@ You can read the following data with this package:
...
+On the `cgi` source type you can also find:
+- Inverter temperature
+
 On the `tcp` source type you can also find:
 
-- Inverter temperature;
+- Inverter temperature
 - Voltage and current for the DC input strings (up to 3)
 - Voltage, current, frequency and power for all AC outputs (also up to 3)
 - Total number of runtime hours.
```
Comment (MarijnS95) [inline README.md:105]: Also remove the dot here please :)

Comment (koyote) [reply README.md]: Done

**Comment by koyote** [comment]
Hi,

Is there anything other change needed to get this PR merged?

**Pull request closed without merging**